yktracec/src/jitmodbuilder.cc

```diff
@@ -1912,6 +1922,54 @@ class JITModBuilder {
     finalise(AOTMod, &Builder);
     return JITMod;
   }
+
+  /// Handle promoted variables.
+  void handlePromote(CallInst *CI, MappableFrame *MPF, BasicBlock *BB,
+                     size_t CurBBIdx, size_t CurInstrIdx) {
+    // First lookup the constant value the trace is going to use.
+    uint64_t PConst = __yk_lookup_promote_usize();
+    Value *PConstLL =
+        Builder.getIntN(PointerSizedIntTy->getIntegerBitWidth(), PConst);
+    Value *PromoteVar = CI->getArgOperand(0);
+    Value *JITPromoteVar = getMappedValue(PromoteVar);
+
+    // Now emit a guard for the value, so that we deoptimise if the
+    // value ever changes.
+    MPF->LastSMCall = cast<CallInst>(CI->getNextNonDebugInstruction());
+
+    // We need a way to independently reference both the pre-promoted variable
+    // and the newly promoted constant in our mappings. We acheive this is a
+    // slightly hacky way by making a dummy instruction that we can map to the
+    // pre-promoted variable.
+    Instruction *NewInst = SelectInst::Create(
```

Comment:
Maybe something like:
```
Even though the call `y = yk_promote(x)` is not copied over into the trace and `y` is being replaced with a constant,  we need keep hold of the original value of `y` so it can be accessed during deoptimisation. The simplest way is by inserting a dummy instruction that will be copied into the trace, creating a mapping back to its AOT value.
```

This is not great either (this is difficult to describe), but maybe you can merge this with your comment somehow?

Reply:
I believe this is resolved @ptersilie ?

Reply:
Works for me.